error using variable int to index array containing reference to function (can use int literals) possibly related #16366

### Describe the bug

Trying to execute functions stored in array fails if variable used to index into the array.  No issues using integer literal. 

### Expected Behavior

output from calling functions the stored in the array
```
gday
hallo
gday
aloha

gday
hallo
gday
aloha
```

### Current Behavior

```
~/Develop/vlang/v/v run "/Users/devros/code/vlang/w1dec/arrfntst.v"
arrfntst.v:14:1: warning: unused variable: `tmp`
   12 | println('')
   13 | 
   14 | tmp := 1
      | ~~~
   15 | println( greet[tmp]() ) 
   16 |
arrfntst.v:17:5: warning: unused variable: `ind`
   15 | println( greet[tmp]() ) 
   16 | 
   17 | for ind in 0..greet.len { 
      |     ~~~
   18 |     println( greet[ind]() )
   19 | }
arrfntst.v:15:10: error: unknown function: greet
   13 | 
   14 | tmp := 1
   15 | println( greet[tmp]() ) 
      |          ~~~~~~~~~~~~
   16 | 
   17 | for ind in 0..greet.len {
arrfntst.v:18:11: error: unknown function: greet
   16 | 
   17 | for ind in 0..greet.len { 
   18 |     println( greet[ind]() )
      |              ~~~~~~~~~~~~
   19 | }
```

### Reproduction Steps

```v
type Sfn = fn() string
const greeting = 'hallo gday aloha'.split(' ')

mut greet := []Sfn{ len: 3, init: fn [it] ()string { return greeting[it] } }

println( greet[1]() )

for g in greet {
	println( g() )
}

println('')

// compiler error unknown function greet
tmp := 1
println( greet[tmp]() ) 

for ind in 0..greet.len { 
	println( greet[ind]() )
}


```

### Possible Solution

_No response_

### Additional Information/Context

Possibly related to report #16366 (by andreweathan) using nested map.  
```v
fn main() {
	mut smap := map[string]fn (){}
	mut dmap := map[string]map[string]fn (){}

	smap['one'] = fn () {
		println('One')
	}
	dmap['one']['de'] = fn () {
		println('Eine')
	}

	smap['one']()

	dmap['one']['de']() //throw compile error
	// temp := dmap['one']['de']  // using a temp var works
	// temp()
}
```
```
└> ~/Develop/vlang/v/v run "/Users/devros/code/vlang/w1dec/maperr.v"
==================
   ^~~~~~~~~~~~~~~
/tmp/v_501/maperr.10945530213683240146.tmp.c:12063:63: error: too many arguments provided to function-like macro invocation
        ((void (*) ())(*(voidptr*)map_get(ADDR(map, ADDR(map, dmap), &(string[]){_SLIT("one")}, &(voidptr[]){ new_map(sizeof(string), sizeof(anon_fn_), &map_hash_string, &map_eq_string, &map_clone_string, &map_free_string) })))), &(string[]){_SLIT("de")}, &(voidptr[]){ 0 })))();
                                                                     ^
/tmp/v_501/maperr.10945530213683240146.tmp.c:178:9: note: macro 'ADDR' defined here
#define ADDR(type, expr) (&((type[]){expr}[0]))
        ^
/tmp/v_501/maperr.10945530213683240146.tmp.c:12063:36: error: use of undeclared identifier 'ADDR'
        ((void (*) ())(*(voidptr*)map_get(ADDR(map, ADDR(map, dmap), &(string[]){_SLIT("one")}, &(voidptr[]){ new_map(sizeof(string), sizeof(anon_fn_), &map_hash_string, &map_eq_string, &map_clone_string, &map_free_string) })))), &(string[]){_SLIT("de")}, &(voidptr[]){ 0 })))();
                                          ^
1 warning and 2 errors generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

```

### V version

V 0.3.2 5acd855

### Environment details (OS name and version, etc.)

OS: macos, macOS, 13.0.1, 22A400 Processor: 8 cpus, 64bit, little endian, Apple M2 CC version: Apple clang version 14.0.0 (clang-1400.0.29.202)

--- a/vlib/v/parser/parser.v
+++ b/vlib/v/parser/parser.v
@@ -2272,6 +2272,11 @@ fn (p &Parser) is_generic_call() bool {
 					if nested_sbr_count > 0 {
 						nested_sbr_count--
 					} else {
+						prev_tok := p.peek_token(i - 1)
+						// `funcs[i]()` is not generic call
+						if !(p.is_typename(prev_tok) || prev_tok.kind == .rsbr) {
+							return false
+						}
 						if p.peek_token(i + 1).kind == .lpar {
 							return true
 						}

--- a/vlib/v/tests/array_of_functions_direct_call_test.v
+++ b/vlib/v/tests/array_of_functions_direct_call_test.v
@@ -1,0 +1,20 @@
+fn foo() string {
+	println('foo')
+	return 'foo'
+}
+
+fn bar() string {
+	println('bar')
+	return 'bar'
+}
+
+fn call() {
+	funcs := [foo, bar]
+	i := 1
+	ret := funcs[i]()
+	assert ret == 'bar'
+}
+
+fn test_array_of_functions_direct_call() {
+	call()
+}

